accord-core/src/main/java/accord/utils/async/AsyncChainCombiner.java

```diff
@@ -0,0 +1,168 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package accord.utils.async;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+
+import java.util.List;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+
+abstract class AsyncChainCombiner<I, O> extends AsyncChains.Head<O>
+{
+    private static final AtomicIntegerFieldUpdater<AsyncChainCombiner> REMAINING = AtomicIntegerFieldUpdater.newUpdater(AsyncChainCombiner.class, "remaining");
+    private volatile Object state;
+    private volatile BiConsumer<? super O, Throwable> callback;
+    private volatile int remaining;
+
+    protected AsyncChainCombiner(List<AsyncChain<I>> inputs)
+    {
+        Preconditions.checkArgument(!inputs.isEmpty());
+        this.state = inputs;
+    }
+
+    private List<AsyncChain<? extends I>> inputs()
+    {
+        Object current = state;
+        Preconditions.checkState(current instanceof List);
+        return (List<AsyncChain<? extends I>>) current;
+    }
+
+    private I[] results()
+    {
+        Object current = state;
+        Preconditions.checkState(current instanceof Object[]);
+        return (I[]) current;
+    }
+
+    void add(AsyncChain<I> chain)
+    {
+        inputs().add(chain);
+    }
+
+    void addAll(List<AsyncChain<I>> chains)
+    {
+        inputs().addAll(chains);
+    }
+
+    int size()
+    {
+        Object current = state;
+        if (current instanceof List)
+            return ((List) current).size();
+        if (current instanceof Object[])
+            return ((Object[]) current).length;
+        throw new IllegalStateException();
+    }
+
+    abstract void complete(I[] results, BiConsumer<? super O, Throwable> callback);
+
+    private void callback(int idx, I result, Throwable throwable)
+    {
+        int current = remaining;
+        if (current == 0)
+            return;
+
+        if (throwable != null && REMAINING.compareAndSet(this, current, 0))
+        {
+            callback.accept(null, throwable);
+            return;
+        }
+
+        results()[idx] = result;
+        if (REMAINING.decrementAndGet(this) == 0)
+        {
+            try
+            {
+                complete(results(), callback);
+            }
+            catch (Throwable t)
+            {
+                callback.accept(null, t);
```

Comment:
I don't think this is safe, as we may be catching an exception thrown by the callback itself in `complete`. I think we need to say `complete` takes control of the callback, and invokes it exceptionally if it is unable to invoke it unexceptionally. But once invoked, control is handed to the callback, and we shouldn't invoke it _again_ with an exception.